pkg/account/authorization.go

```diff
@@ -0,0 +1,185 @@
+package account
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/cloudtrust/keycloak-bridge/internal/dto"
+
+	cs "github.com/cloudtrust/common-service"
+	"github.com/cloudtrust/common-service/security"
+
+	"github.com/cloudtrust/common-service/log"
+	api "github.com/cloudtrust/keycloak-bridge/api/account"
+)
+
+// Creates constants for API method names
+const (
+	UpdatePassword            = "UpdatePassword"
+	GetCredentials            = "GetCredentials"
+	GetCredentialRegistrators = "GetCredentialRegistrators"
+	UpdateLabelCredential     = "UpdateLabelCredential"
+	DeleteCredential          = "DeleteCredential"
+	MoveCredential            = "MoveCredential"
+	GetAccount                = "GetAccount"
+	UpdateAccount             = "UpdateAccount"
+	DeleteAccount             = "DeleteAccount"
+	GetConfiguration          = "GetConfiguration"
+)
+
+// Tracking middleware at component level.
+type authorizationComponentMW struct {
+	logger         log.Logger
+	configDBModule ConfigurationDBModule
+	next           Component
+}
+
+// MakeAuthorizationAccountComponentMW checks authorization and return an error if the action is not allowed.
+func MakeAuthorizationAccountComponentMW(logger log.Logger, configDBModule ConfigurationDBModule) func(Component) Component {
+	return func(next Component) Component {
+		return &authorizationComponentMW{
+			logger:         logger,
+			configDBModule: configDBModule,
+			next:           next,
+		}
+	}
+}
+
+// authorizationComponentMW implements Component.
+func (c *authorizationComponentMW) UpdatePassword(ctx context.Context, currentPassword, newPassword, confirmPassword string) error {
+	var action = UpdatePassword
+	var currentRealm = ctx.Value(cs.CtContextRealm).(string)
+
+	var config = dto.RealmConfiguration{}
+	var err error
+
+	if config, err = c.configDBModule.GetConfiguration(ctx, currentRealm); err != nil {
+		infos, _ := json.Marshal(map[string]string{
+			"currentRealm": currentRealm,
+		})
+		c.logger.Error("Error", "Configuration not found", "infos", string(infos))
+		return err
+	}
+
+	if !*config.APISelfPasswordChangeEnabled {
+		infos, _ := json.Marshal(map[string]string{
+			"Action":       action,
+			"currentRealm": currentRealm,
+		})
+		c.logger.Debug("ForbiddenError", "Password change disabled", "infos", string(infos))
```

Comment:
If we use infos rather than string(infos), Logger.Debug() won't do the job anyway ?

Reply:
If we don't use string(), we have a base64 string